File: js/2623.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const [N, M] = inputs[0].split(' ').map(Number);

const adjList = Array(N + 1)
  .fill()
  .map(() => []);

const inDegree = Array(N + 1).fill(0);

for (let i = 1; i <= M; i++) {
  const arr = inputs[i].split(' ').map(Number);
  const n = arr[0];
  for (let i = 1; i < n; i++) {
    adjList[arr[i]].push(arr[i + 1]);
    inDegree[arr[i + 1]]++;
  }
}

const ans = [];

let idx = 0;
const q = [];
for (let i = 1; i <= N; i++) {
  if (inDegree[i] === 0) q.push(i);
}

while (q.length !== idx) {
  const n = q[idx++];
  ans.push(n);

  for (let i = 0; i < adjList[n].length; i++) {
    const nn = adjList[n][i];
    inDegree[nn]--;
    if (inDegree[nn] === 0) q.push(nn);
  }
}

console.log(ans.length !== N ? 0 : ans.join('\n'));

File: c++/2623.cpp
// bfs
// #include <iostream>
// #include <queue>
// #include <vector>

// using namespace std;

// vector<int> adjList[1001];
// int indegree[1001];

// int main() {
//   freopen("example.txt", "r", stdin);

//   int N, M;
//   cin >> N >> M;

//   while (M--) {
//     int n;
//     cin >> n;
//     vector<int> v;
//     while (n--) {
//       int a;
//       cin >> a;
//       v.push_back(a);
//     }

//     for (int i = 0; i < v.size() - 1; i++) {
//       adjList[v[i]].push_back(v[i + 1]);
//       indegree[v[i + 1]]++;
//     }
//   }

//   queue<int> q;
//   vector<int> ans;

//   for (int i = 1; i <= N; i++) {
//     if (indegree[i] == 0) q.push(i);
//   }

//   while (!q.empty()) {
//     int n = q.front();
//     q.pop();
//     ans.push_back(n);

//     for (int i = 0; i < adjList[n].size(); i++) {
//       int nn = adjList[n][i];

//       indegree[nn]--;
//       if (indegree[nn] == 0) {
//         q.push(nn);
//       }
//     }
//   }

//   if (ans.size() != N) {
//     cout << 0;
//     return 0;
//   }

//   for (int i = 0; i < ans.size(); i++) {
//     cout << ans[i] << '\n';
//   }

//   return 0;
// }

// dfs
#include <iostream>
#include <vector>

using namespace std;

vector<int> adjList[1001];
vector<int> stack;
bool isVisited[1001];
bool finish[1001];
bool isCycle = false;

void dfs(int n) {
  isVisited[n] = true;

  for (int i = 0; i < adjList[n].size(); i++) {
    int nn = adjList[n][i];
    if (!isVisited[nn]) {
      dfs(nn);
    } else if (!finish[nn]) {
      isCycle = true;
    }
  }
  finish[n] = true;
  stack.push_back(n);
}

int main() {
  freopen("example.txt", "r", stdin);

  int N, M;
  cin >> N >> M;

  while (M--) {
    int n;
    cin >> n;
    vector<int> v;
    while (n--) {
      int a;
      cin >> a;
      v.push_back(a);
    }

    for (int i = 0; i < v.size() - 1; i++) {
      adjList[v[i]].push_back(v[i + 1]);
    }
  }
  for (int i = 1; i <= N; i++) {
    if (!isVisited[i]) {
      dfs(i);
    }
  }

  if (isCycle) {
    cout << 0;
    return 0;
  }

  for (int i = stack.size() - 1; i >= 0; i--) {
    cout << stack[i] << '\n';
  }

  return 0;
}